File: Fortran/04_One_sided/03_Accumulate/accumulate.f90
PROGRAM accumulate

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100

  INTEGER :: rank, src, dest, nproc
  INTEGER :: ierr, window, intsize
  INTEGER(KIND=MPI_ADDRESS_KIND) :: size_of_window, offset

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !In C would just use sizeof()
  !In Fortran2008, could use C_SIZEOF
  !NOTE! This routine returns the size of AN ELEMENT
  !Not the size of an array if you give it an array
  CALL MPI_Sizeof(src, intsize, ierr)
  !Just using a single int here
  size_of_window = intsize

  !Create the window. This is a piece of memory that's available for remote
  !access. In this case, a single 4 byte integer
  CALL MPI_Win_create(dest, size_of_window, intsize, MPI_INFO_NULL, &
      MPI_COMM_WORLD, window, ierr)

  !Use collective synchronization model. After this command any processor
  !can use MPI_Put or MPI_Get on any other processor
  CALL MPI_Win_fence(0, window, ierr)

  src = rank
  dest = 0

  !Put the result into the first (zeroth) slot
  offset = 0

  !This call accumulates whatever data is in src on all processors and puts
  !them in the window on processor 0. In this demo, we are using MPI_Win_fence
  !so everything can both send and receive. In theory, rank 0 needs to be
  !in both the access epoch and the exposure epoch, and the other ranks
  !only need to be in the exposure epoch
  CALL MPI_Accumulate(src, , , , , , , &
      , , )

  !Call Win_fence again to end the access and exposure epochs
  CALL MPI_Win_fence(0, window, ierr)

  IF (rank == 0) PRINT *,"Accumulated value is ", dest

  CALL MPI_Finalize(ierr)

END PROGRAM accumulate

File: C/04_One_sided/03_Accumulate/accumulate.c
#include <stdio.h>
#include <mpi.h>

#define TAG 100

int main(int argc, char ** argv)
{

  int rank, nproc, src, dest;
  MPI_Aint size_of_window, offset;
  MPI_Win window;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  offset = 0;
  size_of_window = sizeof(int);

  //Create the window. This is a piece of memory that's available for remote
  //access. In this case, a single 4 byte integer
  MPI_Win_create(&dest, size_of_window, sizeof(int), MPI_INFO_NULL,
      MPI_COMM_WORLD, &window);

  //Use collective synchronization model. After this command any processor
  //can use MPI_Put or MPI_Get on any other processor
  MPI_Win_fence(0,window);

  //Put the result into the first (zeroth) slot
  offset = 0;

  //Accumulate ranks and zero destination
  src = rank;
  dest = 0;

  MPI_Accumulate(&src, , , , , , ,
      , );

  //Call Win_fence again to end the access and exposure epochs
  MPI_Win_fence(0, window);

  if (rank == 0) printf("Accumulated value is %3d\n", dest);


  MPI_Finalize();

}
